juju/model.py

```diff
@@ -2116,7 +2115,7 @@ async def _deploy(self, charm_url, application, series, config,
                 devices=devices,
                 dryrun=False,
                 placement=placement,
-                storage={k: parse_storage_constraint(v) for k, v in (storage or dict()).items()},
+                storage=storage,
```

Comment:
We should add validation that the storage value passed to `_deploy()` is of the expected format. Outside of the test for bundle processing, there's very little if any test coverage for that calls to `model.deploy()` validate input that I am aware of. You could image it would be easy for the unparsed storage string "ebs,1G,2" etc to get to the wrong place instead of the parsed value. We do have unit tests for `parse_storage_constraint` but I wonder if we should include a couple of different variations of the storage string besides the fully populated '"ebs,100G,1"` in the bundle tests.

Reply:
Yeah I saw that there were already unit tests for `constraint.py` so I just grabbed a valid value out of the docs to make the integration tests pass -- previously these tests just checked that the string "storage" made it through to the mock deploy unmodified. I considered using several different values here, but since the purpose of these tests isn't to check that the storage constraint parsing works for different values, as that's validated separately in unit tests, I opted not to

In terms of runtime validation, `model.deploy` expects `storage` to be a dict like {"label": {count: 1, pool: "pool", size: 1234}}. It tries to unpack `storage['label']` with the `**` operator, so passing a dict like {"label": "ebs,1G,2"} will fail at runtime with a `TypeError`, which luckily is an appropriate error here. After this, the types of the `count`, `pool`, and `size` values will be validated by `juju.client._definitions.Constraints`.

I'll add some extra unit tests to specifically cover variations on the storage args for `bundle.AddApplicationChange`. Maybe we can open a separate issue about test coverage / runtime validation for `model.deploy`.